issue-2282: Client changes

shewitt-au opened the pull request:
<!--
Please provide as much information as possible about what your PR aims to do.
PRs with no description will most likely be closed until more information is provided.
If you're planing on changing fundamental behaviour or add big new features, please open a GitHub Issue first before starting to work on it.
If it's not something big and you still want to contact us about it, feel free to do so !
-->

### Problem description
<!-- Describe the bug that you fixed/feature request that you implemented, or link to an existing issue describing it -->

### Implementation description
<!-- Explain what you did to correct the problem -->

### Screenshots
<!-- If your change is visual, take a screenshot showing it. Ideally, make before/after sceenshots -->

### Additional things
<!-- Anything else you would like to say -->

paxcut commented:
I tested the msi installer and it does seem to fix the problem. I ran the test case that created the crash and now I can both open the nodes and hover over the array without crashes or issues

shewitt-au commented:
Ok. I have updated both PRs.

I rolled back a lot of the code I was working on as it was getting too error prone. I have added a [utility](https://github.com/shewitt-au/std-enable_shared_from_this-and-constructors) to pattern_language to make the two stage creation easier and to allows gradual/on-demand adoption: `lib\include\pl\helpersconstruct_shared_object.hpp`

I have not patched any of the patterns to use two stage creation yet.

Another issue is I suspect we’ll get `std::shared_ptr` cycles, and thus memory leaks, because the parent pointer is a` std::shared_ptr` and one can only assume the parent has `std::shared_ptr`'s to its children. I think the parent pointer should be a weak_ptr.

Builds fine for me. The client side GitHub tests fail however. Not sure why. Perphaps it's building without the matching pattern_language changes?

shewitt-au commented:
OK. Seems good for me so far. I've added enough two-stage constructions to fix all the bugs I've encountered. I’m sure there’s more. A framework is (hopefully) in place to easily fix issues as they arise.

paxcut commented:
I wanted to test your changes but your builds are failing. The build includes both the ImHex and pattern language changes and the first error can be traced bask to this [line](https://github.com/WerWolv/PatternLanguage/blob/3ec6ae20ac2b5ded5829eefaa660f29b67ff1229/lib/source/pl/core/evaluator.cpp#L355) in your Imhex repository.
```cpp
    std::shared_ptr<construct_shared_object<pl::ptrn::Pattern> Evaluator>Variable(const std::string &name, const ast::ASTNodeTypeDecl *type, const std::optional<Token::Literal> &value, bool outVariable, bool reference, bool templateVariable, bool constant) {
 ```
 I am not sure where that comes from but if you follow the links in your Imhex repository for the pattern language linked there that what you'll find in line 355. I don't see any commits that have that change so you must be linking some other commit as a submodule.

shewitt-au commented:
@paxcut 
I put in a fix. Tests running now. I'm confused though. "so you must be linking some other commit as a submodule." How do I check for that?

shewitt-au commented:
@paxcut 

Failed again. Had another crack. Tests running. They are not fail so quickly. Looks like you were're spot on. Thanks for pointing out the root cause. With my git knowledge I would not have fixured that out.

shewitt-au commented:
Ok. Only tests failed, and they don't like they are caused by my changes.
